File: Lab2/CSharp/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharp {
	public static partial class Ext {
		public static IEnumerable<int> MergeSort(this int[] source) {
			if (source.IsEmpty()) {
				return source;
			}
			if (source.Length == 1) {
				return source;
			}
			var splittedArrays = source.SplitInto(2);
			var left = MergeSort(splittedArrays[0]);
			var right = MergeSort(splittedArrays[1]);
			return Merge(left, right);
		}

		internal static IEnumerable<int> Merge(this IEnumerable<int> left, IEnumerable<int> right) {
			if (!left.Any()) {
				return right;
			}
			if (!right.Any()) {
				return left;
			}
			var leftHead = left.First();
			var leftTail = left.Skip(1);
			var rightHead = right.First();
			var rightTail = right.Skip(1);
			if (leftHead < rightHead) {
				return leftHead.ToEnumerable().Concat(Merge(leftTail, right));
			} else {
				return rightHead.ToEnumerable().Concat(Merge(left, rightTail));
			}
		}

		internal static T[][] SplitInto<T>(this T[] source, int count) {
			IEnumerable<T[]> SplitInto() {
				int length = source.Length;
				count = length < count ? length : count;
				if (count == 0) {
					yield return Array.Empty<T>();
				}
				int maxPartLength = (int)Math.Ceiling((double)length / count);
				for (int i = 0; i < count; i++) {
					int sourceIndex = i * maxPartLength;
					var diff = length - sourceIndex - maxPartLength;
					int partLength = diff >= 0 ? maxPartLength : length - sourceIndex;
					T[] part = new T[partLength];
					Array.Copy(source, sourceIndex, part, 0, partLength);
					yield return part;
				}
			};
			return SplitInto().ToArray();
		}
	}
}

File: Lab2/FSharp/MergeSort.fs
module MergeSort

let rec merge_v1List left right = 
            match left with
            | [] -> right
            | leftHead::leftTail -> 
                match right with
                | [] -> left
                | rightHead::rightTail ->
                    if leftHead < rightHead then leftHead::(merge_v1List leftTail right) else rightHead::(merge_v1List left rightTail)

let rec merge_v2Seq left right = 
            match (left, right) with
            | (seq, _) when Seq.isEmpty seq -> right
            | (_, seq) when Seq.isEmpty seq -> left
            | (leftSeq, rightSeq) ->                 
                let leftHead = Seq.head leftSeq
                let leftTail = Seq.tail leftSeq
                let rightHead = Seq.head rightSeq
                let rightTail = Seq.tail rightSeq
                if leftHead < rightHead 
                    then Seq.append (Seq.singleton leftHead) (merge_v2Seq leftTail rightSeq)
                    else Seq.append (Seq.singleton rightHead) (merge_v2Seq left rightTail)

let rec mergeSort_v1List (array: 'a array) = 
    match array with
    | [||] -> Array.toList array
    | [|x|] -> Array.toList array
    | _ -> 
        let splitSortToList = Array.splitInto 2 >> Array.map mergeSort_v1List >> Array.map Seq.toList
        let [| left; right |] = splitSortToList array        
        merge_v1List left right
   
let rec mergeSort_v2Seq (array: 'a array) = 
    match array with
    | [||] -> Seq.empty
    | [|x|] -> Seq.singleton x
    | _ -> 
        let splitSort = Array.splitInto 2 >> Array.map mergeSort_v2Seq
        let [| left; right |] = splitSort array        
        merge_v2Seq left right